### ISSUE TITLE
Make links clickable for talkback users

### ISSUE THREAD

**Issue by schulle4u** [andstatus/andstatus#392]
When using talkback screen-reader for android, links in andstatus cannot be opened. Sweeping left or right with the finger only sets the focus to the whole message, but not to parts like links or hashtags. The only way to access links is to open the message (tweet etc) from context menu in a browser and open the actual link from there. 
Using Android 5.1 / Talkback 4.5.1. 

Steffen

**Comment by yvolk** [comment]
Hi @schulle4u , I installed Google Talkback and tried AndStatus with it.
Yes, I see that links are not clickable by the tool.
I also found similar question http://stackoverflow.com/questions/38444822/how-to-make-links-among-other-text-focusable-for-accessibility with an answer that having links inside a textbox clickable for accessibility is not supported by the Android Platform.

I hear that each clickable link is preceded by a chime, but that's all.

I could create some special layout for accessibility, in which each "span" would have its own textbox. So some textboxes could be clickable...

Do you have any good example for what you need, preferably on Android? Or maybe you have a vision/know, how this could be implemented?

**Comment by schulle4u** [comment]
Hello @yvolk, 

Sorry, I'm not an android developer, my technical insight therefore is very limited. Apps which I know of to work with embedded links are the K9 mailclient and its message view, as well as the crome browser and apps based upon android system web view. 

Building an extra accessibility design seems not the best solution to me, since it would take also extra time for you to maintain this whenever something have to be changed. 
My idea for now is to add an appropriate possibility to the context menu of a tweet. There already is an option to open the whole tweet directly on twitter. Could another option be added, telling the application to open one or more links attached to the tweet? By far not the best solution, but this wouldn't requiere you to rewrite the whole interface for just this one thing. 

Steffen

**Comment by yvolk** [comment]
Hello Steffen,
So we are talking about another message ContextMenu, which would contain a list of links from this message?!
Maybe additional button could be added to the button bar below each message. Click on that button will open the list of links.
Or you would prefer to turn that "context" buttons off altogether... ?!

**Comment by schulle4u** [comment]
@yvolk: I personally prefer the buttons to get out of my way and use the context menu if needed, since I mostly use AndStatus for reading messages then for interacting with them. But if I remember right, the context menu provides identical functionality as the buttons, so a "link list" button/submenu could be added for both scenarios. 

Steffen

**Comment by yvolk** [comment]
I got it, so:
In a case "the touch exploration in the system is enabled" AndStatus adds new item to the top of the message's context menu: "X message links", where "X" is actual number of clickable links. Clicking on this item opens the submenu with an item for each link.

?!

See the coding examples: http://www.programcreek.com/java-api-examples/index.php?source_dir=alarm-clock-master/src/com/klinker/deskclock/widget/multiwaveview/GlowPadView.java

I also found this interesting and heated discussion on "[Programmatically checking for Accessibility - if TalkBack is on](https://groups.google.com/forum/#!topic/eyes-free/9pajGQLp4X0)"

**Comment by schulle4u** [comment]
Sounds like a good plan Yuri, I'll look forward to the next releases on F-Droid. Thanks in advance for having an eye on this topic. :)

Steffen

**Comment by yvolk** [comment]
Implemented, will be available since v.30,01
This is how it looks now. 
1. A User navigates in the list, reading messages using a Talkback application
![scr-touch-exploration](https://cloud.githubusercontent.com/assets/207716/19281253/0e628af4-8ff2-11e6-8037-c4b62cf0b943.png)

2. Double click on a message to "activate" it, i.e. to open its context menu:
![scr-n-message-links](https://cloud.githubusercontent.com/assets/207716/19281275/2982cc18-8ff2-11e6-944e-5fd4bdecba97.png)

3. Select and then double-click on the "2 message links" menu item. Submenu opens. Now clicking on any link will open a browser window.
![scr-n-message-links-submenu](https://cloud.githubusercontent.com/assets/207716/19281280/3404173c-8ff2-11e6-83bd-233cd536fbc5.png)

**Comment by schulle4u** [comment]
Many thanks, this works perfectly.

**Comment by rabid01** [comment]
hi your code is not available on the given link... please share it

**Comment by yvolk** [comment]
@rabid01 What exactly link doesn't work?

**Comment by rabid01** [comment]
http://www.programcreek.com/java-api-examples/index.php?source_dir=alarm-clock-master/src/com/klinker/deskclock/widget/multiwaveview/GlowPadView.java  this link is not showing your code

**Comment by yvolk** [comment]
@rabid01 That was not my code, but some example from Internet that I found at that time.
My application is still here, at GitHub.
